arangod/Graph/Enumerators/WeightedTwoSidedEnumerator.h

```diff
@@ -121,12 +203,37 @@ class WeightedTwoSidedEnumerator {
     ProviderType _provider;
 
     PathValidatorType _validator;
-    std::unordered_map<typename Step::VertexType, size_t> _visitedNodes;
+    std::unordered_map<typename Step::VertexType, std::vector<size_t>>
+        _visitedNodes;
     Direction _direction;
     GraphOptions _graphOptions;
   };
   enum BallSearchLocation { LEFT, RIGHT, FINISH };
 
+  /*
+   * A class that is able to store valid shortest paths results.
+   * This is required to be able to check for duplicate paths, as those can be
+   * found during a KShortestPaths graphs search.
+   */
+  class ResultCache {
+   public:
+    ResultCache(Ball& left, Ball& right);
+    ~ResultCache();
+
+    // @brief: returns whether a path could be inserted or not.
+    // True: It will be inserted if this specific path has not been added yet.
+    // False: Could not insert as this path has already been found.
+    auto addResult(CalculatedCandidate const& candidate) -> bool;
```

Comment:
```suggestion
    [[nodiscard]] auto tryAddResult(CalculatedCandidate candidate) -> bool;
```
you could pass by value here and use `std::move()` at the call site (assuming you are not using the candidate further), or we could even enforce a move by using `&&`

Reply:
No, we do need the candidate. We cannot move it. 
e.g. it is moved directly afterwards if valid: 
```
if (foundNonDuplicatePath) {
          handleCandidate(std::move(potentialCandidate));
          return true;
        }
```

Reply:
Still added nodiscard and changed name to "tryAddResult".